blockchain/consensus/narwhal_cns/src/exec.rs

```diff
@@ -0,0 +1,174 @@
+            .is_ok()
+    }
+
+    fn release_consensus_write_lock(&self) {
+        self.locked.store(false, Ordering::Relaxed)
+    }
+}
+
+#[async_trait]
+impl<BS> BatchExecutionState for NarwhalExecutionState<BS>
+where
+    BS: BlockStore + Sync + Send + 'static,
+{
+    /// Load the tip of the chain and get the last certificate index from the `Ticket` in the header.
+    async fn load_next_certificate_index(&self) -> Result<SequenceNumber, Self::Error> {
+        match self.chain_store.heaviest_tipset().await {
+            None => Ok(0),
+            Some(tipset) => {
+                let index = ConsensusTransactionIndex::try_from(tipset.as_ref())?;
+                Ok(index.next_consensus_index())
+            }
+        }
+    }
+
+    async fn handle_consensus(
+        &self,
+        consensus_output: &ConsensusOutput,
+        transaction_batches: Vec<Vec<SerializedTransaction>>,
+    ) -> Result<(), Self::Error> {
+        // Append the transactions and the certificate to a pending queue that we can take values from
+        // when we have extended the local blockchain with a block. It is tempting to try to keep the
+        // last block we built here and just keep appending on top of it, or to append it directly to
+        // the chainstore. But that would bypass all the regular channels in `TipsetProcessor` and
+        // `ChainMuxer`, which are quite difficult to grasp, making this a risky option.
+
+        let output = NarwhalOutput {
+            consensus_output: consensus_output.clone(),
+            transaction_batches,
+        };
+        self.output_tx
+            .send(output)
+            .await
+            .map_err(|_| NarwhalConsensusError::Other("Cannot send output.".into()))?;
+        Ok(())
+    }
+}
+
+/// Support crash recovery by storing the index of processed transactions in the block tickets.
+/// Looking at the these numbers we can decide if we need to replay from the current, or the next consensus index.
+#[derive(Serialize_tuple, Deserialize_tuple, Clone, Debug)]
+pub struct ConsensusTransactionIndex {
+    /// The consensus index the transactions in the block came from.
+    pub consensus_index: SequenceNumber,
+    /// The number of transactions included in blocks (this and previous ones) from this consensus output so far.
+    pub transactions_included: u32,
+    /// The total number of transactions in this consensus output.
+    pub transactions_total: u32,
+}
+
+impl ConsensusTransactionIndex {
+    fn next_consensus_index(&self) -> SequenceNumber {
+        if self.transactions_included < self.transactions_total {
+            self.consensus_index
+        } else {
+            self.consensus_index + 1
+        }
+    }
+}
+
+impl Cbor for ConsensusTransactionIndex {}
+
+/// Parse the certificate index from the `Ticket` in a `Tipset`.
+///
+/// While storing the index in the `Ticket` is a bit of a hack, this way we can be sure that storing
```

Comment:
Agree! We use the same approach of "abusing" the `Ticket` in other consensus algorithms.

Reply:
Yeah, that's where it came from :relaxed: 
I'm don't even think our current database abstractions allow atomic reads/writes of multiple operations, so I didn't think I could write a block and then make sure the node doesn't die before saving the consensus index as well.
I have a whole set of abstractions in one of my repos to do just that, for a rainy day :)